blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md

```diff
@@ -0,0 +1,294 @@
+reconcile this by considering *enriched* category theory and the
+*enriched* coYoneda lemma.  Let \\(V \\) be a category, then if \\( V
+\\) is sufficiently like the category of sets, then we can do a lot of
+category theory by replacing the word "set" with "object of \\(V \\)".
+
+Specifically, a \\(V \\)-enriched category (or just \\(V\\)-category)
+has a set of objects \\(Ob \\), but for each pair of objects \\(A,B
+\in Ob \\) we get a \\( V\\)-object \\(\Hom(A,B) \\) of morphisms from
+\\(A \\) to \\( B \\).  If \\(V \\) is a closed category, we can see
+\\(V \\) *itself* as a \\(V\\)-enriched category with the same objects
+and just making \\(\Hom(A,B) = A \to B \\).
+
+Then we can reinterpret the coYoneda lemma above by saying \\(C \\) is
+a \\(V\\)-category and \\(Q \\) is a \\(V\\)-presheaf i.e., just a
+contravariant functor from \\(V \\) to itself: \\(Q : V^{op} \to V\\)
+where the preprocessing function is now a morphism in \\(C \\).
+Haskelletons just call this
+a
+[contravariant functor](https://hackage.haskell.org/package/contravariant-1.4/docs/Data-Functor-Contravariant.html).
+Furthermore, since existential types provide at least as strong of a
+reasoning principle as coends, the proof of the coYoneda lemma goes
+through with existential types instead.
+Finally, the point-free description above for coend can be interpreted
+in any category.
+
+Now that we're working all inside our language, let's look at what the
+isomorphism looks like in Haskellish/Agdaish syntax.
+We want mutually inverse functions
+
+```
+f : (Contravariant Q) => (∃ Γ. (Δ -> Γ) × (Q Γ)) -> Q Δ
+g : (Contravariant Q) => Q Δ -> ∃ Γ. (Δ -> Γ) × (Q Γ)
+```
+
+If you try to implement them you won't be able to get it wrong, but here they are:
+
+```
+f (k, qΓ) = contramap k qΓ
+g qΔ = (id, qΔ)
+```
+
+where we just instantiate \\(\Gamma = \Delta \\) in the second case.
+You can prove \\( f \circ g = id \\) using just \\(\beta \\) and the
+Contravariant laws, but to prove \\(g \circ f = id \\) you need to use
+the coend reasoning.
+For those of you that know about the Yoneda lemma, note the similarity
+to that proof in using the identity function and instantiating a type
+variable in a trivial way.
+
+## Closure Version as CoYoneda
+
+Now it's time to bring it all together.
+Let \\(V \\) be our programming language viewed as a category in the
+usual way.
+
+We want to prove the closure conversion isomorphism:
+
+\\[ A \to B \cong \exists \Gamma. \Gamma \times (\Gamma \times A \to B)
+\\]
+
+using the \\(V \\)-coYoneda lemma which says for any contravariant
+functor \\(Q : V^{op} \to V \\),
+
+\\[ Q(\Delta) \cong \exists \Gamma. (\Delta \to \Gamma) \times Q(\Gamma)
+\\]
+
+Clearly based on the right hand side, \\(Q \\) should be \\( - \times
+A \to B \\) which gives us:
+
+\\[ \Delta \times A \to B \cong \exists \Gamma. (\Delta \to \Gamma) \times (\Gamma \times A \to B)
+\\]
+
+and if we pick \\(\Delta = 1\\), the unit type and the fact that \\(1
```

Comment:
this sentence is too long,

suggestion: "Let `\Delta = 1`. Since `1 \times A \cong A` and `1 \to \Gamma \cong \Gamma` (at least ....), we get the desired result ...."